modules/auxiliary/gather/ldap_esc_vulnerable_cert_finder.rb

```diff
@@ -0,0 +1,544 @@
+        '(pkiextendedkeyusage=1.3.6.1.5.5.7.3.2)'\
+        '(pkiextendedkeyusage=1.3.6.1.5.2.3.4)'\
+        '(pkiextendedkeyusage=2.5.29.37.0)'\
+        '(!(pkiextendedkeyusage=*))'\
+      ')'\
+      '(mspki-certificate-name-flag:1.2.840.113556.1.4.804:=1)'\
+    ')'
+    attributes = ['cn', 'description', 'ntSecurityDescriptor']
+    base_prefix = 'CN=Certificate Templates,CN=Public Key Services,CN=Services,CN=Configuration'
+    esc1_entries = query_ldap_server(esc1_raw_filter, attributes, base_prefix: base_prefix)
+    fail_with(Failure::NotVulnerable, 'Could not find any ESC1 vulnerable certificate templates!') if esc1_entries.blank?
+
+    # Grab a list of certificates that contain vulnerable settings.
+    # Also print out the list of SIDs that can enroll in that server.
+
+    esc1_entries.each do |entry|
+      @vuln_certificate_template_list << entry[:cn][0]
+
+      begin
+        security_descriptor = Rex::Proto::MsDtyp::MsDtypSecurityDescriptor.read(entry[:ntsecuritydescriptor][0])
+      rescue IOError => e
+        fail_with(Failure::UnexpectedReply, "Unable to read security descriptor! Error was: #{e.message}")
+      end
+
+      flag_allowed_to_enroll, allowed_sids = parse_dacl_or_sacl(security_descriptor.dacl) if security_descriptor.dacl
+      if flag_allowed_to_enroll == true
+        @vuln_cert_template_table << ['ESC1', entry[:cn][0], entry[:dn][0], convert_sids_to_human_readable_name(allowed_sids)]
+      end
+      parse_dacl_or_sacl(security_descriptor.sacl) if security_descriptor.sacl
+    end
+  end
+
+  def find_esc2_vuln_cert_templates
+    esc2_raw_filter = '(&'\
+      '(objectclass=pkicertificatetemplate)'\
+      '(!(mspki-enrollment-flag:1.2.840.113556.1.4.804:=2))'\
+      '(|(mspki-ra-signature=0)(!(mspki-ra-signature=*)))'\
+      '(|'\
+        '(pkiextendedkeyusage=2.5.29.37.0)'\
+        '(!(pkiextendedkeyusage=*))'\
+      ')'\
+    ')'
+
+    attributes = ['cn', 'description', 'ntSecurityDescriptor']
+    base_prefix = 'CN=Certificate Templates,CN=Public Key Services,CN=Services,CN=Configuration'
+    esc2_entries = query_ldap_server(esc2_raw_filter, attributes, base_prefix: base_prefix)
+    fail_with(Failure::NotVulnerable, 'Could not find any ESC2 vulnerable certificate templates!') if esc2_entries.blank?
+
+    # Grab a list of certificates that contain vulnerable settings.
+    # Also print out the list of SIDs that can enroll in that server.
+
+    esc2_entries.each do |entry|
+      @vuln_certificate_template_list << entry[:cn][0]
+
+      begin
+        security_descriptor = Rex::Proto::MsDtyp::MsDtypSecurityDescriptor.read(entry[:ntsecuritydescriptor][0])
+      rescue IOError => e
+        fail_with(Failure::UnexpectedReply, "Unable to read security descriptor! Error was: #{e.message}")
+      end
+
+      flag_allowed_to_enroll, allowed_sids = parse_dacl_or_sacl(security_descriptor.dacl) if security_descriptor.dacl
+      if flag_allowed_to_enroll == true
```

Comment:
There a reason this needs to be explicitly checked against `true`?
```suggestion
      if flag_allowed_to_enroll
```

Reply:
No particular reason I think I just was trying to compare it to a particular value but reality is that it doesn't matter.

Reply:
Sorry I'm unclear where this is being set to `false` in the code, do you have some examples?

Reply:
It's not set to false, but it needs to be on each iteration or you need to update the logic. It's only set to the return value of `parse_dacl_or_sacl` if `security_descriptor.dacl` is truthy. So if you process one entry, `security_descriptor.dacl` is truthy, you call `parse_dacl_or_sacl` and set `flag_allowed_to_enroll` to true, then process another entry and `security_descriptor.dacl` is false then `flag_allowed_to_enroll` is left unmodified (so it'll be true from the previous iteration) causing a false positive.

See this comment where I make suggestions of how to fix it.
https://github.com/rapid7/metasploit-framework/pull/17122/files#r1010688254

Reply:
Removed unneeded True comparisons with https://github.com/rapid7/metasploit-framework/pull/17122/commits/0814ecd98262358e9fca1bc9ae159d5195fa4bf6

Reply:
Fixed the second issue with `flag_allowed_to_enroll` with https://github.com/rapid7/metasploit-framework/pull/17122/commits/fd33cdc58e47576823a874c5f17d81b3e5660f9b